pkg/build/controller/policy/policy.go

```diff
@@ -0,0 +1,116 @@
+package policy
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	buildclient "github.com/openshift/origin/pkg/build/client"
+	buildutil "github.com/openshift/origin/pkg/build/util"
+	"k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+)
+
+// RunPolicy is an interface that define handler for the build runPolicy field.
+// The run policy controls how and when the new builds are 'run'.
+type RunPolicy interface {
+	// IsRunnable returns true of the given build should be executed.
+	IsRunnable(*buildapi.Build) (bool, error)
+
+	// OnComplete allows policy to execute action when the given build just
+	// completed.
+	OnComplete(*buildapi.Build) error
+}
+
+// GetDefaultRunPolicies returns a set of all default run policies.
+func GetDefaultRunPolicies(lister buildclient.BuildLister, updater buildclient.BuildUpdater) []RunPolicy {
+	return []RunPolicy{
+		&ParallelPolicy{BuildLister: lister, BuildUpdater: updater},
+		&SerialPolicy{BuildLister: lister, BuildUpdater: updater},
+		&SerialLatestOnlyPolicy{BuildLister: lister, BuildUpdater: updater},
+	}
+}
+
+// ForBuild picks the appropriate run policy for the given build.
+func ForBuild(build *buildapi.Build, policies []RunPolicy) RunPolicy {
+	for _, s := range policies {
+		switch buildutil.BuildRunPolicy(build) {
+		case buildapi.BuildRunPolicyParallel:
+			if _, ok := s.(*ParallelPolicy); ok {
+				glog.V(5).Infof("Using %T run policy for build %s/%s", s, build.Namespace, build.Name)
+				return s
+			}
+		case buildapi.BuildRunPolicySerial:
+			if _, ok := s.(*SerialPolicy); ok {
+				glog.V(5).Infof("Using %T run policy for build %s/%s", s, build.Namespace, build.Name)
+				return s
+			}
+		case buildapi.BuildRunPolicySerialLatestOnly:
+			if _, ok := s.(*SerialLatestOnlyPolicy); ok {
+				glog.V(5).Infof("Using %T run policy for build %s/%s", s, build.Namespace, build.Name)
+				return s
+			}
+		}
+	}
+	return nil
+}
+
+// GetNextConfigBuild returns the build that will be executed next for the given
+// build configuration. It also returns the indication whether there are
+// currently running builds, to make sure there is no race-condition between
+// re-listing the builds.
+func GetNextConfigBuild(lister buildclient.BuildLister, namespace, buildConfigName string) (*buildapi.Build, bool, error) {
+	var (
+		nextBuild           *buildapi.Build
+		hasRunningBuilds    bool
+		previousBuildNumber int64
+	)
+	builds, err := buildutil.BuildConfigBuilds(lister, namespace, buildConfigName, func(b buildapi.Build) bool {
+		switch b.Status.Phase {
+		case buildapi.BuildPhasePending, buildapi.BuildPhaseRunning:
+			hasRunningBuilds = true
+			return false
+		}
```

Comment:
can't you also leave out any build that's in a completed/canceled/failed state?  in fact i think you should, otherwise you might update its start time later.